cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java

```diff
@@ -43,6 +46,38 @@ public WorkflowSpecification(String className, String name, String description,
     this.properties = properties == null ? Collections.<String, String>emptyMap() :
                                            Collections.unmodifiableMap(new HashMap<>(properties));
     this.nodes = Collections.unmodifiableList(new ArrayList<>(nodes));
+    this.nodeIdMap = Collections.unmodifiableMap(generateNodeIdMap(nodes));
+  }
+  /**
+   +   * Visit all the nodes in the {@link Workflow} and generate the map of node id to
+   +   * the {@link WorkflowNode}.
+   +   */
+  private Map<String, WorkflowNode> generateNodeIdMap(List<WorkflowNode> nodesInWorkflow) {
```

Comment:
It has been copied and is needed for this PR. If that PR gets committed before this, will remove my changes